quarkc/lib/promises.q

```diff
@@ -0,0 +1,268 @@
+quark *;
+include io/datawire/quark/runtime/Lock.java;
+
+namespace quark {
+namespace promises {
+
+    @doc("A callable that takes a single argument, returns result.")
+    interface UnaryCallable {
+      Object invoke(Object arg);
+    }
+
+    // Called when Promise we're waiting on has result, allowing us to hand it
+    // over to Promise that is waiting for it:
+    class _ChainPromise extends UnaryCallable {
+        Promise _next;
+
+        _ChangePromise(Promise next) {
+            self._next = next;
+        }
+
+        Object invoke(Object arg) {
+            _CallbackEvent.fullfilPromise(self.next, arg);
+        }
+    }
+
+    // Actually run a callback's callable, and hand result on to promise that is
+    // expecting it:
+    class _CallbackEvent extends concurrent.Event {
+        UnaryCallable _callable;
+        Promise _next;
+        Object _value;
+
+        _CallbackEvent(UnaryCallable callable, Promise next, Object value) {
+            self._callable = callable;
+            self._next = next;
+            self._value = value;
+        }
+
+        static void fullfilPromise(Promise promise) {
+            if (reflect.Class.ERROR.hasInstance(self._value)) {
+                promise._reject(self._value);
+            } else {
+                promise._resolve(self._value);
+            }
+        }
+
+        void fireEvent() {
+            Object result = self._callable.invoke(self._value);
+            if (reflect.Class.get("quark.promises.Promise").hasInstance(result)) {
+                // We got a promise as result of callback, so chain it to the
+                // promise that we're supposed to be fulfilling:
+                Promise toChain = result;
+                result.then(new _ChainPromise(self._next));
+            } else {
+                self.fullfilPromise(self._next);
```

Comment:
... so why do we fulfill the promise with `_value` and not with `result` in contrast with the case the function returned a `Promise`